contracts/NonfungiblePositionManagerV4.sol

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+import {IERC721Metadata} from "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol";
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {IERC721} from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";
+
+import {INonfungiblePositionManagerV4} from "./interfaces/INonfungiblePositionManagerV4.sol";
+import {PeripheryValidation} from "./base/PeripheryValidation.sol";
+import {PeripheryPayments} from "./base/PeripheryPayments.sol";
+import {PeripheryImmutableState} from "./base/PeripheryImmutableState.sol";
+import {SelfPermit} from "./base/SelfPermit.sol";
+import {LiquidityManagement} from "./base/LiquidityManagement.sol";
+import {Multicall} from "./base/Multicall.sol";
+
+contract NonfungiblePositionManagerV4 is
+    INonfungiblePositionManagerV4,
+    ERC721,
+    PeripheryImmutableState,
+    PeripheryValidation,
+    LiquidityManagement,
+    SelfPermit,
+    Multicall
+{
+    using PoolIdLibrary for PoolKey;
+
+    error InvalidTokenID();
+    error NotApproved();
+    error NotCleared();
+    error NonexistentToken();
+
+    // details about the Uniswap position
+    struct TokenPosition {
+        // the nonce for permits
+        uint96 nonce;
+        // the address that is approved for spending this token
+        address operator;
+        // the hashed poolKey of the pool with which this token is connected
+        PoolId poolId;
+        // the tick range of the position
+        int24 tickLower;
+        int24 tickUpper;
+        // the liquidity of the position
+        uint128 liquidity;
+        // the fee growth of the aggregate position as of the last action on the individual position
+        uint256 feeGrowthInside0LastX128;
+        uint256 feeGrowthInside1LastX128;
+        // how many uncollected tokens are owed to the position, as of the last computation
+        uint128 tokensOwed0;
+        uint128 tokensOwed1;
+    }
+
+    /// @dev Pool keys by poolIds
+    mapping(bytes32 => PoolKey) private _poolIdToPoolKey;
+
+    /// @dev The token ID position data
+    mapping(uint256 => TokenPosition) private _positions;
```

Comment:
thinking they probably want to use the `positions` getter for getting position data, since it includes helpers for unwrapping `poolId` - otherwise they're kind of exposed to our internal representation of how we store pools right? they'll need both `_poolIdToPoolKey` and `_positions` to get useful position data

Reply:
oh yeah good point

Reply:
nit: Would just add to the natspec that they key is specifically the tokenId (tokenId) ie mapping from tokenId to position data